pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,337 @@
+	groupAddress []string
+	event        ruleEvent
+}
+
+// multicastController handles only IGMP protocol. It stores the group address to RuleID map and
+// also rule to group addresses map
+type multicastController struct {
+	ofClient openflow.Client
+
+	ifaceStore interfacestore.InterfaceStore
+	ruleCache  *ruleCache
+	queue      workqueue.RateLimitingInterface
+	// key is the multicast group address.
+	mcastItemRuleCache cache.Indexer
+
+	// key is ruleID while the value is list of group addresses.
+	ruleIDGroupAddressMap map[string]sets.String
+	ruleIDGroupMapMutex   sync.RWMutex
+}
+
+func groupAddressKey(obj interface{}) (string, error) {
+	group := obj.(*mcastItem)
+	return group.groupAddress.IP.String(), nil
+}
+
+func groupAddressRuleIndexFunc(obj interface{}) ([]string, error) {
+	item := obj.(*mcastItem)
+	ruleIDs := make([]string, len(item.ruleIDs))
+	for ruleID := range item.ruleIDs {
+		ruleIDs = append(ruleIDs, ruleID)
+	}
+	return ruleIDs, nil
+}
+
+func (c *multicastController) addGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleAdd,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) updateGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleUpdate,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) deleteGroupAddressForTableIDs(ruleID string, groupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		groupAddress: groupAddresses,
+		event:        ruleDelete,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) validate(iface *interfacestore.InterfaceConfig,
+	groupAddress net.IP, direction crdv1beta.Direction) (*v1alpha1.RuleAction, apitypes.UID, *crdv1beta.NetworkPolicyType, string, error) {
+
+	var ruleTypePtr *crdv1beta.NetworkPolicyType
+	action, uuid, ruleName := v1alpha1.RuleActionAllow, apitypes.UID(""), ""
+	if iface == nil {
+		// Check the if iface exists
+		klog.ErrorS(fmt.Errorf("iface should not be empty"), "")
+		return nil, "", nil, "", fmt.Errorf("iface should not be empty")
+	}
+	ns, podname := iface.PodNamespace, iface.PodName
+	obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress.String())
+	if !exists {
+		klog.V(2).InfoS("Rule for group does not exist", "group", groupAddress.String())
+		action = v1alpha1.RuleActionAllow
+		return &action, "", nil, "", nil
+	}
+	item := obj.(*mcastItem)
+	var matchedRule *CompletedRule
+	member := &crdv1beta.GroupMember{
+		Pod: &crdv1beta.PodReference{
+			Name:      podname,
+			Namespace: ns,
+		},
+	}
+	for ruleID := range item.ruleIDs {
+		// Iterate all rules relate to this multicast group address
+		r, _, _ := c.ruleCache.GetCompletedRule(ruleID)
```

Comment:
I doubt whether this is really necessary to query the ruleCache given reconciler already passes the snapshot of related rules to this controller. It should build a more efficient data struct to achieve the purpose, otherwise the whole controller is not even needed and the validator can just use the ruleCache. But I haven't found how this controller is really used as currently it's not consumed by any clients, will see if this can be improved after you correct the caller side code.

Reply:
How about cache the IGMP rule here, and add indexer for groupAddress? Or add indexer to ruleCache.rules?
Then we can use index to get rules which relates the the groupAddress.